File: Paintroid/src/androidTest/java/org/catrobat/paintroid/test/espresso/util/wrappers/TransformToolOptionsViewInteraction.java
/*
 * Paintroid: An image manipulation application for Android.
 * Copyright (C) 2010-2015 The Catrobat Team
 * (<http://developer.catrobat.org/credits>)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.catrobat.paintroid.test.espresso.util.wrappers;

import android.view.View;
import android.widget.EditText;
import android.widget.TextView;

import org.catrobat.paintroid.R;
import org.hamcrest.Description;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeMatcher;

import static org.catrobat.paintroid.test.espresso.util.UiInteractions.setProgress;
import static org.hamcrest.Matchers.not;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.action.ViewActions.replaceText;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;

public final class TransformToolOptionsViewInteraction extends CustomViewInteraction {
	private TransformToolOptionsViewInteraction() {
		super(onView(withId(R.id.pocketpaint_layout_tool_options)));
	}

	public static TransformToolOptionsViewInteraction onTransformToolOptionsView() {
		return new TransformToolOptionsViewInteraction();
	}

	public TransformToolOptionsViewInteraction performSetCenterClick() {
		onView(withId(R.id.pocketpaint_transform_set_center_btn))
				.perform(click());
		return this;
	}

	public TransformToolOptionsViewInteraction performAutoCrop() {
		onView(withId(R.id.pocketpaint_transform_auto_crop_btn))
				.perform(click());
		return this;
	}

	public TransformToolOptionsViewInteraction checkAutoDisplayed() {
		onView(withText(R.string.transform_auto_crop_text))
				.check(matches(isDisplayed()));
		return this;
	}

	public TransformToolOptionsViewInteraction performRotateClockwise() {
		onView(withId(R.id.pocketpaint_transform_rotate_right_btn))
				.perform(click());
		return this;
	}

	public TransformToolOptionsViewInteraction performRotateCounterClockwise() {
		onView(withId(R.id.pocketpaint_transform_rotate_left_btn))
				.perform(click());
		return this;
	}

	public TransformToolOptionsViewInteraction performFlipVertical() {
		onView(withId(R.id.pocketpaint_transform_flip_vertical_btn))
				.perform(click());
		return this;
	}

	public TransformToolOptionsViewInteraction performFlipHorizontal() {
		onView(withId(R.id.pocketpaint_transform_flip_horizontal_btn))
				.perform(click());
		return this;
	}

	public TransformToolOptionsViewInteraction moveSliderTo(int moveTo) {
		onView(withId(R.id.pocketpaint_transform_resize_seekbar))
				.perform(setProgress(moveTo));
		return this;
	}

	public TransformToolOptionsViewInteraction performApplyResize() {
		onView(withId(R.id.pocketpaint_transform_apply_resize_btn))
				.perform(click());
		return this;
	}

	public TransformToolOptionsViewInteraction performEditResizeTextField(String size) {
		onView(withId(R.id.pocketpaint_transform_resize_percentage_text))
				.perform(replaceText(size));
		return this;
	}

	public TransformToolOptionsViewInteraction checkPercentageTextMatches(int expected) {
		onView(withId(R.id.pocketpaint_transform_resize_percentage_text))
				.check(matches(withText(Integer.toString(expected))));
		return this;
	}

	public TransformToolOptionsViewInteraction checkLayerWidthMatches(int expected) {
		Integer expectedValue = expected;
		onView(withId(R.id.pocketpaint_transform_width_value))
				.check(matches(hasValueEqualTo(expectedValue.toString())));
		return this;
	}

	Matcher<View> hasValueEqualTo(final String content) {

		return new TypeSafeMatcher<View>() {

			@Override
			public void describeTo(Description description) {
				description.appendText("Has EditText/TextView the value:  " + content);
			}

			@Override
			public boolean matchesSafely(View view) {
				if (!(view instanceof TextView) && !(view instanceof EditText)) {
					return false;
				}
				if (view != null) {
					String text;
					if (view instanceof TextView) {
						text = ((TextView) view).getText().toString();
					} else {
						text = ((EditText) view).getText().toString();
					}
					return text.matches(content);
				}
				return false;
			}
		};
	}

	public TransformToolOptionsViewInteraction checkLayerHeightMatches(int expected) {
		Integer expectedValue = expected;
		onView(withId(R.id.pocketpaint_transform_height_value))
				.check(matches(hasValueEqualTo(expectedValue.toString())));
		return this;
	}

	public TransformToolOptionsViewInteraction checkIsDisplayed() {
		onView(withId(R.id.pocketpaint_layout_tool_options))
				.check(matches(isDisplayed()));
		return this;
	}

	public TransformToolOptionsViewInteraction checkIsNotDisplayed() {
		onView(withId(R.id.pocketpaint_layout_tool_options))
				.check(matches(not(isDisplayed())));
		return this;
	}
}

File: Paintroid/src/androidTest/java/org/catrobat/paintroid/test/espresso/util/wrappers/TransformToolOptionsViewInteraction.kt
/*
 * Paintroid: An image manipulation application for Android.
 * Copyright (C) 2010-2022 The Catrobat Team
 * (<http://developer.catrobat.org/credits>)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.catrobat.paintroid.test.espresso.util.wrappers

import android.view.View
import android.widget.EditText
import android.widget.TextView
import androidx.test.espresso.Espresso
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.assertion.ViewAssertions
import androidx.test.espresso.matcher.ViewMatchers
import androidx.test.espresso.matcher.ViewMatchers.withId
import androidx.test.espresso.matcher.ViewMatchers.withText
import org.catrobat.paintroid.R
import org.catrobat.paintroid.test.espresso.util.UiInteractions
import org.hamcrest.Description
import org.hamcrest.Matchers
import org.hamcrest.TypeSafeMatcher

class TransformToolOptionsViewInteraction private constructor() :
    CustomViewInteraction(Espresso.onView(withId(R.id.pocketpaint_layout_tool_options))) {
    fun performSetCenterClick(): TransformToolOptionsViewInteraction {
        Espresso.onView(withId(R.id.pocketpaint_transform_set_center_btn))
            .perform(ViewActions.click())
        return this
    }

    fun performAutoCrop(): TransformToolOptionsViewInteraction {
        Espresso.onView(withId(R.id.pocketpaint_transform_auto_crop_btn))
            .perform(ViewActions.click())
        return this
    }

    fun checkAutoDisplayed(): TransformToolOptionsViewInteraction {
        Espresso.onView(withText(R.string.transform_auto_crop_text))
            .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        return this
    }

    fun performRotateClockwise(): TransformToolOptionsViewInteraction {
        Espresso.onView(withId(R.id.pocketpaint_transform_rotate_right_btn))
            .perform(ViewActions.click())
        return this
    }

    fun performRotateCounterClockwise(): TransformToolOptionsViewInteraction {
        Espresso.onView(withId(R.id.pocketpaint_transform_rotate_left_btn))
            .perform(ViewActions.click())
        return this
    }

    fun performFlipVertical(): TransformToolOptionsViewInteraction {
        Espresso.onView(withId(R.id.pocketpaint_transform_flip_vertical_btn))
            .perform(ViewActions.click())
        return this
    }

    fun moveSliderTo(moveTo: Int): TransformToolOptionsViewInteraction {
        Espresso.onView(withId(R.id.pocketpaint_transform_resize_seekbar))
            .perform(UiInteractions.setProgress(moveTo))
        return this
    }

    fun performApplyResize(): TransformToolOptionsViewInteraction {
        Espresso.onView(withId(R.id.pocketpaint_transform_apply_resize_btn))
            .perform(ViewActions.click())
        return this
    }

    fun performEditResizeTextField(size: String?): TransformToolOptionsViewInteraction {
        Espresso.onView(withId(R.id.pocketpaint_transform_resize_percentage_text))
            .perform(ViewActions.replaceText(size))
        return this
    }

    fun checkPercentageTextMatches(expected: Int): TransformToolOptionsViewInteraction {
        Espresso.onView(withId(R.id.pocketpaint_transform_resize_percentage_text))
            .check(ViewAssertions.matches(withText(expected.toString())))
        return this
    }

    fun checkLayerWidthMatches(expected: Int): TransformToolOptionsViewInteraction {
        Espresso.onView(withId(R.id.pocketpaint_transform_width_value))
            .check(ViewAssertions.matches(hasValueEqualTo(expected.toString())))
        return this
    }

    private fun hasValueEqualTo(content: String): TypeSafeMatcher<View?> {
        return object : TypeSafeMatcher<View?>() {
            override fun describeTo(description: Description) {
                description.appendText("Has EditText/TextView the value:  $content")
            }

            public override fun matchesSafely(view: View?): Boolean {
                if (view !is TextView && view !is EditText) { return false }
                if (view is EditText) { return view.text.toString() == content }
                return (view as TextView).text.toString() == content
            }
        }
    }

    fun checkLayerHeightMatches(expected: Int): TransformToolOptionsViewInteraction {
        Espresso.onView(withId(R.id.pocketpaint_transform_height_value))
            .check(ViewAssertions.matches(hasValueEqualTo(expected.toString())))
        return this
    }

    fun checkIsNotDisplayed(): TransformToolOptionsViewInteraction {
        Espresso.onView(withId(R.id.pocketpaint_layout_tool_options))
            .check(ViewAssertions.matches(Matchers.not(ViewMatchers.isDisplayed())))
        return this
    }

    companion object {
        fun onTransformToolOptionsView(): TransformToolOptionsViewInteraction = TransformToolOptionsViewInteraction()
    }
}
